examples/apps/src/ble_bas_peripheral.rs

```diff
@@ -119,13 +104,22 @@ async fn advertise_task<C: Controller>(
             .await?;
         let conn = advertiser.accept().await?;
         info!("[adv] connection established");
-        // Keep connection alive
         let mut tick: u8 = 0;
-        while conn.is_connected() {
-            Timer::after(Duration::from_secs(2)).await;
-            tick = tick.wrapping_add(1);
-            info!("[adv] notifying connection of tick {}", tick);
-            let _ = server.notify(&server.battery_service.level, &conn, &tick).await;
+        loop {
+            match select(conn.next(), Timer::after(Duration::from_secs(2))).await {
+                Either::First(event) => match event {
+                    ConnectionEvent::Disconnected { reason } => {
+                        info!("[adv] disconnected: {:?}", reason);
+                        break;
+                    }
+                    _ => {}
```

Comment:
https://github.com/jamessizeland/microbit-ble-gamepad/pull/3/files# updated my PR to show how I'd use this new API design. Looks good! My only suggestion would be to make it { event, connection }, so we can { event, .. } out the connection when not used? (I'm assuming event is the more commonly used return variable.)

```rust
pub async fn gatt_server_task(server: &GamepadServer<'_>, conn: &Connection<'_>) {
    // check if the connection is still active every second
    loop {
        if let Either::First(event) = select(conn.next(), server.run()).await {
            match event {
                ConnectionEvent::Disconnected { reason } => {
                    info!("[gatt] Disconnected: {:?}", reason);
                    break;
                }
                ConnectionEvent::Gatt {
                    connection: _,
                    event,
                } => match event {
                    GattEvent::Read { value_handle } => {
                        info!("[gatt] Server Write event on {:?}", value_handle);
                    }
                    GattEvent::Write { value_handle } => {
                        info!("[gatt] Read event on {:?}", value_handle);
                    }
                },
            }
        }
    }
    info!("Gatt server task finished");
}
```

Reply:
You can do { event, ..} independently of the order in the struct. My personal preference is to keep them all in the tests and examples to clearly show what data you get from looking at the code.